modules/navier_stokes/include/userobjects/INSADObjectTracker.h

```diff
@@ -0,0 +1,135 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "GeneralUserObject.h"
+#include "MooseTypes.h"
+
+#include "libmesh/vector_value.h"
+
+template <typename>
+class ADMaterialProperty;
+template <typename>
+class MaterialProperty;
+
+/**
+ * Object for tracking what kernels have been added to an INSAD simulation. This is used to
+ * determine what properties to calculate in the INSADMaterial, which is important particularly for
+ * ensuring that we have consistenly included all the strong terms for stabilization methods like
+ * PSPG and SUPG
+ */
+class INSADObjectTracker : public GeneralUserObject
+{
+public:
+  static InputParameters validParams()
+  {
+    InputParameters params = GeneralUserObject::validParams();
+    params.addClassDescription("User object used to track the kernels added to an INS simulation "
+                               "and determine what properties to calculate in INSADMaterial");
+    return params;
```

Comment:
You might want to set 'execute_on' = INITIAL?

Reply:
That's fair, although the `execute` method is empty